### ISSUE TITLE
very slow translation...

### ISSUE THREAD

**Issue by philippehenri** [ModernMT/MMT#392]
Hello,

To test MMT, I have trained it with 20 Millions sentences. Now when i try a translation of a 40 words sentence, it takes more than 5 seconds…

Is it normal, is there a way to improve this ? any suggestion please?

note: Machine is with 1 geforce GTX 1080, Intel® Core i7-6700 Quadcore processor built using Skylake architecture, 64 GB DDR4 RAM, and 500 GB 6 Gb/s SSDs. Ubuntu 16.04 - I have tried it on another machine with a nvidia P4000 and had the same speed issue...

thanks !
philippe

**Comment by nicolabertoldi** [comment]
Actually, we measured an average translation time below 1 second per sentence.
Have you trained with the default parameters, or have you changed anything?
Which version (commit) of MMT are you using?
Could you please:
- start the engine with the parameter "-v 2"
- translate a bunch of sentences
- stop the engine
- send us the log file  "runtime/default/logs/node.log"

**Comment by philippehenri** [comment]
Hi Nicola

thanks for your reply. Unfortunately it seems that i have to restart as i wanted to do more tests with a smaller dataset and my other memory seems to have disappeared.

Now to answer your questions: I am using 2.5 and no i havent changed anything, just copied two very large parallel files into a directory and i ran:

./mmt create nl fr examples/data/test

I will update you with the logs when training is complete again.

But i have a question in the meantime please:

1/ what is the difference between "neural" and "phrase_based", i thought MMT was all neural translation?

2/ how can i train the same parallel corpus both way "fr-nl" and "nl-fr", do i have to run it twice?

It appears that if i run:

./mmt create nl fr examples/data/test
then
./mmt create fr nl examples/data/test

the second overwrites the first.

thanks
ph

**Comment by nicolabertoldi** [comment]
1/ what is the difference between "neural" and "phrase_based", i thought MMT was all neural translation?
MMT system implemented two decoding framework:
"neural":    which exploits neural machine translation
"phrase_based":    which exploit statistical machine translation

I would kindly suggest you to use "neural", because in the future release "phrase-based" will be dismissed completely

2/ how can i train the same parallel corpus both way "fr-nl" and "nl-fr", do i have to run it twice?
yes, training is monodirectional; hence you have to train the two engines ("fr-nl" and "nl-fr") separately.
By default mmt creates an engine called "default"; hence, the second invocation of the training overwrites the first.
You can easily specify a different name for each engine with the parameter "--engine ENGINE_NAME";  in your case you could run the two commands
./mmt create nl fr examples/data/test  --engine my_engine_nl_fr
./mmt create fr nl examples/data/test  --engine my_engine_fr_nl

./mmt create --help
gives you some more details about training parameters

**Comment by philippehenri** [comment]
ok great. so i guess i was using phrase-based with the slow translation (as i could see that in engine.xconf), i have now restarted training with ./mmt create nl fr examples/data/moni -e nlfr --neural

Will report speed results as soon as training is done

**Comment by philippehenri** [comment]
HI
The training is still running. please see  part of the log below. How long will this training run  to complete (currently running since 24 hours approx)?
is there some place where i could find info about that?
thanks
ph

2018-08-15 06:12:31,653 [INFO] - Validation at step 350000: loss = 0.759968, perplexity = 2.13821, accuracy = 83.2361
2018-08-15 06:12:31,653 [INFO] - Validation perplexity at step 350000 (epoch 3.17): 2.138207; new best: 2.138207
2018-08-15 06:12:31,653 [INFO] - Optimizer learning rate decay de-activated at step 350000 (epoch 3.17); current lr value: 0.478297
2018-08-15 06:12:31,653 [INFO] - Checkpoint at step 350000 (epoch 3.17): [num_correct: 18182331.00; 20634797 src tok; 22279211 tgt tok; acc:  81.61; ppl:   2.31; 10081 src tok/s; 10884 tgt tok/s]
2018-08-15 06:12:32,253 [INFO] - Checkpoint saved: path = /home/mmt/mmt/runtime/nlfr/tmp/training/onmt_model/checkpoint_350000 ppl = 2.14
2018-08-15 06:12:32,254 [INFO] - Terminate policy: avg_ppl = 2.2834, previous_avg_ppl = 2.30018, stopping = False
2018-08-15 06:12:52,493 [INFO] - Step 350100 (epoch: 3.17): [num_correct: 181383.00; 206912 src tok; 222388 tgt tok; acc:  81.56; ppl:   2.32; 8835 src tok/s; 9496 tgt tok/s]
2018-08-15 06:13:14,390 [INFO] - Step 350200 (epoch: 3.17): [num_correct: 193616.00; 220401 src tok; 237496 tgt tok; acc:  81.52; ppl:   2.31; 10065 src tok/s; 10846 tgt tok/s]
2018-08-15 06:13:34,951 [INFO] - Step 350300 (epoch: 3.17): [num_correct: 175847.00; 198656 src tok; 215064 tgt tok; acc:  81.76; ppl:   2.28; 9662 src tok/s; 10460 tgt tok/s]

**Comment by nicolabertoldi** [comment]
Training ends when the perplexity (avg_ppl) does not increase with respect to the previous values (previous_avg_ppl) for several (I think 10) times 
It is difficult to say in advance when it happens, 
you could monitor the trend by looking in the log the line "Terminate policy:”  and make a reasonable guess.

In any case, when the avg_ppl trend seems almost flat, you could interrupt the training
by sending the signal SIGINT  to the process  “mmt create”
i.e.    CTRL-C  if the process is in foreground      or    "kill -2’ if the process is in background


Please, next time I would kindly ask you to open a different issue if you change the focus of your request.


> On 15 Aug 2018, at 09:09, philippehenri <notifications@github.com> wrote:
> 
> HI
> The training is still running. please see part of the log below. How long will this training run to complete (currently running since 24 hours approx)?
> is there some place where i could find info about that?
> thanks
> ph
> 
> 2018-08-15 06:12:31,653 [INFO] - Validation at step 350000: loss = 0.759968, perplexity = 2.13821, accuracy = 83.2361
> 2018-08-15 06:12:31,653 [INFO] - Validation perplexity at step 350000 (epoch 3.17): 2.138207; new best: 2.138207
> 2018-08-15 06:12:31,653 [INFO] - Optimizer learning rate decay de-activated at step 350000 (epoch 3.17); current lr value: 0.478297
> 2018-08-15 06:12:31,653 [INFO] - Checkpoint at step 350000 (epoch 3.17): [num_correct: 18182331.00; 20634797 src tok; 22279211 tgt tok; acc: 81.61; ppl: 2.31; 10081 src tok/s; 10884 tgt tok/s]
> 2018-08-15 06:12:32,253 [INFO] - Checkpoint saved: path = /home/mmt/mmt/runtime/nlfr/tmp/training/onmt_model/checkpoint_350000 ppl = 2.14
> 2018-08-15 06:12:32,254 [INFO] - Terminate policy: avg_ppl = 2.2834, previous_avg_ppl = 2.30018, stopping = False
> 2018-08-15 06:12:52,493 [INFO] - Step 350100 (epoch: 3.17): [num_correct: 181383.00; 206912 src tok; 222388 tgt tok; acc: 81.56; ppl: 2.32; 8835 src tok/s; 9496 tgt tok/s]
> 2018-08-15 06:13:14,390 [INFO] - Step 350200 (epoch: 3.17): [num_correct: 193616.00; 220401 src tok; 237496 tgt tok; acc: 81.52; ppl: 2.31; 10065 src tok/s; 10846 tgt tok/s]
> 2018-08-15 06:13:34,951 [INFO] - Step 350300 (epoch: 3.17): [num_correct: 175847.00; 198656 src tok; 215064 tgt tok; acc: 81.76; ppl: 2.28; 9662 src tok/s; 10460 tgt tok/s]
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 


-- 
--
Le informazioni contenute nella presente comunicazione sono di natura 
privata e come tali sono da considerarsi riservate ed indirizzate 
esclusivamente ai destinatari indicati e per le finalità strettamente 
legate al relativo contenuto. Se avete ricevuto questo messaggio per 
errore, vi preghiamo di eliminarlo e di inviare una comunicazione 
all’indirizzo e-mail del mittente.

--
The information transmitted is 
intended only for the person or entity to which it is addressed and may 
contain confidential and/or privileged material. If you received this in 
error, please contact the sender and delete the material.

**Comment by philippehenri** [comment]
oops sorry Nicola, well noted ! About the speed issue, here is the log you requested yesterday. phrase-based engine (20M), you will note the slow translations (30 secs, 10 secs, ...
[modern mt log.txt](https://github.com/ModernMT/MMT/files/2289528/modern.mt.log.txt)
).

**Comment by nicolabertoldi** [comment]
I do not see any problem in your log,
I also suppose no problem occurred during training of the PB models

- MMT tends to be a bit slower at the beginning because it has to load in memory several object, but then it tends to speedup;
  could you please try to translate tens of sentences and see the speed improves

- with PB engine translation time is somehow proportional to the sentence length;    from our measurement, we were able to translate sentence up to 20 words in less than 1 second
   could you please try to translate shorter sentences (1-20) words and keep track of the translation speed?


- PB engine works on CPU only;
   is the CPU free, or there are other concurrent processes?
   how much RAM is taken from the process?


- MMT exploits multithreading; hence, the overall  time to translate a batch of sentences is definitely lower than that taken to translate the same sentences sequentially.
  In your scenario, do you need actually need a sequential translation or  can you send batch of sentences

**Comment by philippehenri** [comment]
ok thanks, will try to warm it up and hope it will help.

**Comment by nicolabertoldi** [comment]
I’m closing this issue because it has been inactive for more than one week. This probably means that  it has been fixed or solved .
Please feel free to reopen if you still encounter this issue.

Thank you!
